### PR TITLE
Market decentralized/confirm agreement

### LINKED ISSUE #378: Initiate Agreement with provider
- [x] Implement POST `agreement/{agreement_id}/confirm` endpoint
    - [x] Send agreement to provider
    - [x] Provider adds Agreement to database
    - [x] Provider adds event; Agreement is available at query events endpoint

to be continued in #431

### PR DESCRIPTION
**Pull request by tworec**
resolves #378

### CODE REVIEW

File: core/market/decentralized/src/protocol/negotiation/requestor.rs [hunk 1]
```diff
@@ -118,17 +118,12 @@ impl NegotiationApi {
 
     /// Sent to provider, when Requestor will call confirm Agreement.
     /// TODO: send Agreement content
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/src/protocol/negotiation/requestor.rs:120]: Outdated TODO?

File: core/market/decentralized/src/db/dao/agreement.rs [hunk 2]
```diff
@@ -17,6 +20,37 @@ impl<'a> AsDao<'a> for AgreementDao<'a> {
...
+                .optional()?
+            {
+                None => return Ok(None),
+                Some(agreement) => agreement,
+            };
+
+            if agreement.valid_to < validation_ts {
+                agreement.state = AgreementState::Expired;
+                update_state(conn, &id, &agreement.state)?;
+            }
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/src/db/dao/agreement.rs:42]: :+1: But note that you do this in readonly_transaction

File: core/market/decentralized/migrations/2020-05-15-093513_init-db/up.sql [hunk 3]
```diff
@@ -104,6 +104,9 @@ CREATE TABLE market_agreement(
 	offer_properties TEXT NOT NULL,
 	offer_constraints TEXT NOT NULL,
 
+    offer_id VARCHAR(97) NOT NULL,
+    demand_id VARCHAR(97) NOT NULL,
+
 	provider_id VARCHAR(20) NOT NULL,
 	requestor_id VARCHAR(20) NOT NULL,
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/migrations/2020-05-15-093513_init-db/up.sql:112]: Tabs indentation in other fields in market_agreement. (I know, it's my copy-paste fault :P)

File: core/market/decentralized/src/db/dao/agreement.rs [hunk 4]
```diff
@@ -17,6 +20,37 @@ impl<'a> AsDao<'a> for AgreementDao<'a> {
...
+            Ok(Some(agreement))
+        })
+        .await
+    }
+
+    pub async fn update_state(&self, id: &AgreementId, state: AgreementState) -> DbResult<bool> {
+        let id = id.clone();
+        do_with_transaction(self.pool, move |conn| update_state(conn, &id, &state)).await
+    }
+
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/src/db/dao/agreement.rs:53]: Important: We should check state of Agreement, before changing it, in the same transaction. Now we have dangerous pattern: we get Agreement in first transaction, check state and than change it in second transaction.
I don't like this either, that we are forced to implement something-like state machine on dao level :(

File: core/market/decentralized/src/negotiation/requestor.rs [hunk 5]
```diff
@@ -219,9 +233,45 @@ impl RequestorBroker {
...
+            Some(agreement) => agreement,
+        };
+
+        if agreement.state == AgreementState::Proposal {
+            self.api.propose_agreement(agreement).await?;
+            dao.update_state(agreement_id, AgreementState::Pending)
+                .await
+                .map_err(|e| AgreementError::Get(agreement_id.clone(), e))?;
+            return Ok(());
+        }
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/src/negotiation/requestor.rs:263]: Is it safe to call confirm_agreement form 2 concurrent threads?
(I think it won't hurt us, but I want be sure you verified it as well)

File: core/market/decentralized/src/db/dao/agreement.rs (hunk 4 above)
Comment (tworec) [reply core/market/decentralized/src/db/dao/agreement.rs:51]: added TODO

File: core/market/decentralized/src/negotiation/requestor.rs (hunk 5 above)
Comment (tworec) [reply core/market/decentralized/src/negotiation/requestor.rs:234]: moved to #430

**Pull request merged**